src/vs/editor/contrib/codeAction/browser/codeActionUi.ts

```diff
@@ -192,6 +193,8 @@ export class CodeActionUi extends Disposable {
 			}
 		};
 
+		await Promise.all(actionsToShow.map(async (action) => action.resolve(CancellationToken.None)));
```

Comment:
This will eagerly resolve all code actions, which isn't good for performance. Instead we should only resolve the currently selected action in the list. This is similar to how the suggest widget works

Reply:
@mjbvz  I figured out the _"only resolve the currently selected action_" part by adding the onFocus "listener" to the ActionList class

```typescript
this._register(this._list.onDidChangeFocus(e => this.onListFocus(e)));
```

and then using its delegate object (`IActionListDelegate<CodeActionItem>`) to call resolveCodeAction() for the individual CodeActionItem selected at the moment and set `canPreview` accordingly.

```typescript
private onListFocus(e: IListEvent<IActionListItem<T>>): void {
      this._list.domFocus();
      
      if (!e.elements.length) {
          return;
      }
      
      const element = e.elements[0];
      if (element.item && this.focusCondition(element)) {
          element.canPreview = this._delegate.onFocus!(element.item);
      }
}
```

BUT...I'm having trouble figuring out how to "rerender" that CodeActionItem on the list as to update the tooltip text properly. Could you give some code pointer for that? 🥺

Reply:
To renderer the list, try calling `this._list.rerender()`. You can also remove the modified item and then add it back using `this._list.splice` but that may be more work to hook up correctly